docs: C# example project (#492)
windmilleng/tilt.build: docs/example_csharp.md

---
title: "Example: C#"
layout: docs
lang: csharp
---

The best indicator of a healthy development workflow is a short feedback loop.

Kubernetes is a huge wrench in the works.

Let's fix this.

In this example, we're going to take you through a simple “hello world” web server written in C# that uses [ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/?view=aspnetcore-3.1) as a Model View Controller framework.

We'll use Tilt to:

- Run the server on Kubernetes
- Measure the time from a code change to a new process

> - Optimize that time for fast feedback

Reviewer: ```suggestion
- Optimize that time for faster feedback
```
(for copy consistency as per #487)

This particular example server doesn't do much, but it's useful to confirm that Tilt is working as expected in your environment.

All the code is in this repo:

[tilt-example-csharp](https://github.com/windmilleng/tilt-example-csharp)

To skip straight to the fully optimized setup, go to this subdirectory:

[Recommended Tiltfile](https://github.com/windmilleng/tilt-example-csharp/tree/master/3-live-update)

# Step 0: The Simplest Deployment

Our server has only one page, and all it does is serve us an almost entirely static HTML page:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.RazorPages;
using Microsoft.Extensions.Logging;

namespace hello_tilt.Pages
{
    public class IndexModel : PageModel
    {
        private readonly ILogger<IndexModel> _logger;

        public IndexModel(ILogger<IndexModel> logger)
        {
            _logger = logger;
        }

        public void OnGet()
        {

        }
    }
}
```

> To start this server on Kubernetes, we need three configs:

Reviewer: ```suggestion
To start this server on Kubernetes, we need three config files:
```
(for copy consistency as per #487)

1. A [Dockerfile](https://github.com/windmilleng/tilt-example-csharp/blob/master/0-base/hello-tilt/Dockerfile) that builds the image
2. A [Kubernetes deployment](https://github.com/windmilleng/tilt-example-csharp/blob/master/0-base/kubernetes.yaml) that runs the image
3. And finally, a [Tiltfile](https://github.com/windmilleng/tilt-example-csharp/blob/master/0-base/Tiltfile) that ties them together:

```python
docker_build('hello-tilt', './hello-tilt')
k8s_yaml('kubernetes.yaml')
k8s_resource('hello-tilt', port_forwards='8080:80')
```

The first line tells Tilt to build an image with the name `hello-tilt` in the directory `./hello-tilt`.

The second line tells Tilt to load the Kubernetes Deployment YAML. The image name in the `docker_build` call must match the container `image` reference in the `hello-tilt` Deployment.

> The last line configures port-forwarding so that your server is reachable at `localhost:8080`. The resource name in the `k8s_resource` call must match the Deployment's `metadata.name.

Reviewer: ```suggestion
The last line configures port-forwarding so that your server is reachable at `localhost:8080`. The resource name in the `k8s_resource` call must match the Deployment's `metadata.name`.
```

```
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://[::]:
```

{% assign case = "base" %}
{% include example_guide_image.html %}

## Step 1: Let's Add Benchmark Trickery

Before we try to make this faster, let's measure it.

Using [`local_resource`](local_house.html), you can direct Tilt to execute existing scripts or arbitrary shell commands on your own machine, and manage them from your sidebar like any other Tilt resource. We're going to use this functionality to benchmark our deploys.

First we add a `local_resource` to our [Tiltfile](https://github.com/windmilleng/tilt-example-csharp/blob/master/1-measure/Tiltfile) that records the start time in a C# file.

```python
local_resource(
    'deploy',
    './record-start-time.sh',
    deps=['./record-start-time.sh']
)
```

The `local_resource()` call creates a local resource named `deploy`. The second argument is the script that it runs.

We've also modified our server to read that start itme, calculate the time elapsed, then display this in the rendered HTML.

Let's click the button on the `deploy` resource and see what happens!

> <figure>

Reviewer: you should probably use templating for all of your images -- but i'm applying review comments on #488 and how we do this will change slightly, so stand by (or else merge as is and i'll mix as part of my PR)

Tilt offers you some default benchmarks _and_ the tools to capture your own.

Our benchmarks show this is slow. Can we do better?

## Step 2: Let's Optimize for the C# Toolchain

What's taking up so much time? The logs show that when we make a change to a file, we:

1) Copy the csproj file, run `dotnet restore` to install any dependencies
2) Copy the rest of the code and run a build from scratch with `dotnet publish`
3) Copy the build output to the ASP.NET runtime image

But the C# community has done a lot of work to make caching dependendencies and incremental compiles fast. How can we better use the tools how they're meant to be used?

With `local_resource`, we can compile the project locally, and copy the build output files to the container.

Here's our [new Tiltfile](https://github.com/windmilleng/tilt-example-csharp/blob/master/2-build-locally/Tiltfile) with the following new code:

```python
local_resource(
    'build',
    'dotnet publish -c Release -o out',
    deps=['hello-tilt'],
    ignore=['hello-tilt/obj'],
    resource_deps=['deploy'],
)
```

> We've added a `local_resource()` that compiles the executable Jar locally with Gradle.

Reviewer: did you need to adjust the build context for your `docker_build` at all? might be useful to mention--otherwise users will see duplicative builds and not necessarily know how to debug.

Author: Yes! Good call, I'll add a note about that.

Tilt offers you some default benchmarks _and_ the tools to capture your own.

Our benchmarks show this is slow. Can we do better?

## Step 2: Let's Optimize for the C# Toolchain

What's taking up so much time? The logs show that when we make a change to a file, we:

1) Copy the csproj file, run `dotnet restore` to install any dependencies
2) Copy the rest of the code and run a build from scratch with `dotnet publish`
3) Copy the build output to the ASP.NET runtime image

But the C# community has done a lot of work to make caching dependendencies and incremental compiles fast. How can we better use the tools how they're meant to be used?

With `local_resource`, we can compile the project locally, and copy the build output files to the container.

Here's our [new Tiltfile](https://github.com/windmilleng/tilt-example-csharp/blob/master/2-build-locally/Tiltfile) with the following new code:

```python
local_resource(
    'build',
    'dotnet publish -c Release -o out',
    deps=['hello-tilt'],
    ignore=['hello-tilt/obj'],
    resource_deps=['deploy'],
)
```

> We've added a `local_resource()` that compiles the executable Jar locally with Gradle. This is p

Author: :sweat_smile:
